### PR TITLE
Add type checking for floating points

### PR DESCRIPTION
**Pull request by yihozhang**

### CODE REVIEW

**Comment by yihozhang** [comment]
After this support, several floats-related benchmarks are taking a significant amount of time to run, e.g., 3mm-test-benchmark took me 400s, and some never finished.

File: dag_in_context/src/type_analysis.egg [hunk 1]
```diff
@@ -206,148 +218,70 @@
...
-        (HasType e1 (Base (IntT)))
-        (HasType e2 (Base (IntT)))
+;; Operators that have type Type -> Type -> Type
+;; Note we only do this generic matching for binary
+;; operator since there's a lot of them.
+;; In the future we can also extend to other constructs.
+(relation bop-of-type (BinaryOp Type))
+(bop-of-type (Add) (Base (IntT)))
+(bop-of-type (Sub) (Base (IntT)))
+(bop-of-type (Div) (Base (IntT)))
```
Comment (yihozhang) [inline dag_in_context/src/type_analysis.egg:229]: Note that the old type checker does not do type checking for integer division, and after supporting this, pure integer benchmarks like `benchmarks/passing/bril/core/digital-root.bril` blow up.

File: dag_in_context/src/type_analysis.egg [hunk 2]
```diff
@@ -206,148 +218,70 @@
...
+        (ExpectType e1 ty op-str)
+        (ExpectType e2 ty op-str)
       )
       :ruleset type-analysis)  
 
-(rule (
-        (= lhs (Bop (LessThan) e1 e2))
-        (HasType e1 (Base (IntT)))
-        (HasType e2 (Base (IntT)))
+;; Operators that have type Float -> Float -> Type
```
Comment (oflatt) [inline dag_in_context/src/type_analysis.egg:253]: This comment seems wrong

File: dag_in_context/src/type_analysis.egg [hunk 3]
```diff
@@ -35,7 +35,15 @@
         (HasType e actual)
         (!= expected actual) ;; OKAY to compare types for equality because we never union types.
       )
-      ((panic "type mismatch, better error message later :(")) ; once egglog allows non-literal panics
+      ((extract "Expecting expression")
```
Comment (oflatt) [inline dag_in_context/src/type_analysis.egg:38]: nice!

**Review by oflatt — CHANGES_REQUESTED** [review]
Minor comments- lets look into the blowup and try to fix or disable optimizations that are busted

**Pull request merged**